R/logging.R

```diff
@@ -0,0 +1,71 @@
+## Plumber has stages preroute -> postroute -> preserialize -> postserialize
+##
+## We hook up the first bit of logging up against postroute as by that
+## point we have run our filters, which include getting the body read
+## and queries parsed. Without this it is not really possible to get
+## the body as we can't portably read twice from the rook input (see
+## warning in ?seek, which is used internally by rewind())
+##
+## The filter functions do minimal validation so there should be very
+## few requests where logging is dropped here.
+porcelain_log_postroute <- function(logger) {
+  force(logger)
+  function(data, req, res) {
+    logger$info("request %s %s", req$REQUEST_METHOD, req$PATH_INFO)
+    logger_detailed(logger, "trace", req, "request")
+  }
+}
+
+
+porcelain_log_postserialize <- function(logger) {
+  force(logger)
+
+  is_json_error_response <- function(res) {
+    res$status >= 400 &&
+      identical(res$headers[["Content-Type"]], "application/json")
+  }
+
+  function(data, req, res, value) {
+    if (is.raw(res$body)) {
+      size <- length(res$body)
+    } else {
+      size <- nchar(res$body)
+    }
+
+    logger$info(sprintf("response %s %s => %d (%d bytes)",
+                        req$REQUEST_METHOD, req$PATH_INFO, res$status, size))
+
+    if (is_json_error_response(res)) {
+      logger_detailed(logger, "error", req, "error",
```

Comment:
I think we don't currently get the trace from this? Or perhaps it is just in the test setup there is no error trace - is this something we can log too?

Reply:
I think that if the trace is added to the error it will come out here, but we don't have an example with this (because it's really a hintr/rrq/orderly.server thing)